File: ios/ios/Core/Security/SSOProtocol.swift
import CryptoKit
import Foundation
import Security

struct SSOCredentials: Sendable {
    var baseURL: URL
    var token: String
    var privateToken: String?
}

enum SSOProtocol {
    static let callbackScheme = "mobilemoodle"
    private static let maxAge: TimeInterval = 10 * 60

    static func createPassport() -> UInt64 {
        var value: UInt64 = 0
        let status = SecRandomCopyBytes(kSecRandomDefault, MemoryLayout<UInt64>.size, &value)
        return status == errSecSuccess ? value & UInt64.max >> 1 : UInt64(Date().timeIntervalSince1970 * 1_000)
    }

    static func launchURL(config: MoodlePublicConfig, passport: UInt64) throws -> URL {
        let selected = config.launchURL.flatMap { MoodleURL.isAllowed(baseURL: config.canonicalURL, candidate: $0) ? $0 : nil }
            ?? MoodleURL.endpoint("admin/tool/mobile/launch.php", at: config.canonicalURL)
        guard var components = URLComponents(url: selected, resolvingAgainstBaseURL: false) else {
            throw MoodleError(code: "invalid_sso", message: String(localized: "error.sso.invalid"))
        }
        components.queryItems = (components.queryItems ?? []) + [
            URLQueryItem(name: "service", value: "moodle_mobile_app"),
            URLQueryItem(name: "passport", value: String(passport)),
            URLQueryItem(name: "urlscheme", value: callbackScheme),
        ]
        guard let result = components.url else { throw MoodleError(code: "invalid_sso", message: String(localized: "error.sso.invalid")) }
        return result
    }

    static func validate(callback: URL, pending: PendingSSO, now: Date = Date()) throws -> SSOCredentials {
        guard callback.scheme == callbackScheme,
              callback.absoluteString.lowercased().hasPrefix("\(callbackScheme)://token=")
        else { throw MoodleError(code: "invalid_sso", message: String(localized: "error.sso.callback"), isRecoverable: false) }
        guard now.timeIntervalSince(pending.createdAt) >= 0, now.timeIntervalSince(pending.createdAt) <= maxAge else {
            throw MoodleError(code: "expired_sso", message: String(localized: "error.sso.expired"), isRecoverable: false)
        }
        let prefix = "\(callbackScheme)://token="
        let encodedPayload = callback.absoluteString.dropFirst(prefix.count).split(whereSeparator: { $0 == "#" || $0 == "?" }).first.map(String.init) ?? ""
        let payload = encodedPayload.removingPercentEncoding ?? encodedPayload
        let parts = payload.components(separatedBy: ":::")
        guard parts.count >= 2, !parts[1].isEmpty else {
            throw MoodleError(code: "invalid_sso", message: String(localized: "error.sso.payload"), isRecoverable: false)
        }
        let expected = Insecure.MD5.hash(data: Data((pending.baseURL.absoluteString + String(pending.passport)).utf8))
            .map { String(format: "%02x", $0) }.joined()
        guard constantTimeEqual(parts[0], expected) else {
            throw MoodleError(code: "invalid_sso", message: String(localized: "error.sso.signature"), isRecoverable: false)
        }
        return SSOCredentials(baseURL: try MoodleURL.normalize(pending.baseURL.absoluteString), token: parts[1], privateToken: parts.count > 2 && !parts[2].isEmpty ? parts[2] : nil)
    }

    private static func constantTimeEqual(_ left: String, _ right: String) -> Bool {
        let lhs = Array(left.utf8), rhs = Array(right.utf8)
        guard lhs.count == rhs.count else { return false }
        return zip(lhs, rhs).reduce(UInt8(0)) { $0 | ($1.0 ^ $1.1) } == 0
    }
}

File: android/app/src/main/java/org/moodle/core/security/SsoProtocol.kt
package org.moodle.core.security

import android.net.Uri
import org.moodle.core.network.MoodleUrl
import java.security.MessageDigest
import java.security.SecureRandom

data class PendingSso(
    val baseUrl: String,
    val passport: Long,
    val createdAtEpochSeconds: Long,
)

data class SsoCredentials(
    val baseUrl: String,
    val token: String,
    val privateToken: String?,
)

object SsoProtocol {
    const val CALLBACK_SCHEME = "mobilemoodle"
    private const val MAX_AGE_SECONDS = 10 * 60

    fun createPassport(): Long = SecureRandom().nextLong().ushr(1)

    fun launchUrl(
        baseUrl: String,
        launchUrl: String?,
        passport: Long,
        service: String = "moodle_mobile_app",
    ): String {
        val normalized = MoodleUrl.normalize(baseUrl)
        val target = launchUrl?.takeIf { MoodleUrl.sameSite(normalized, it) }
            ?: "$normalized/admin/tool/mobile/launch.php"
        return Uri.parse(target).buildUpon()
            .appendQueryParameter("service", service)
            .appendQueryParameter("passport", passport.toString())
            .appendQueryParameter("urlscheme", CALLBACK_SCHEME)
            .build()
            .toString()
    }

    fun validateCallback(
        callback: Uri,
        pending: PendingSso,
        nowEpochSeconds: Long,
    ): SsoCredentials {
        require(callback.scheme == CALLBACK_SCHEME && callback.schemeSpecificPart.startsWith("//token=")) {
            "Unexpected SSO callback"
        }
        require(nowEpochSeconds - pending.createdAtEpochSeconds in 0..MAX_AGE_SECONDS) {
            "The SSO request expired"
        }

        val payload = callback.schemeSpecificPart.substringAfter("//token=", "")
            .substringBefore('#')
            .substringBefore('?')
        val parts = payload.split(":::")
        require(parts.size >= 2 && parts[1].isNotBlank()) { "Invalid SSO payload" }
        require(constantTimeEquals(parts[0], md5(pending.baseUrl + pending.passport))) {
            "Invalid SSO signature"
        }

        return SsoCredentials(
            baseUrl = MoodleUrl.normalize(pending.baseUrl),
            token = parts[1],
            privateToken = parts.getOrNull(2)?.takeIf(String::isNotBlank),
        )
    }

    private fun md5(value: String): String = MessageDigest.getInstance("MD5")
        .digest(value.toByteArray(Charsets.US_ASCII))
        .joinToString("") { "%02x".format(it) }

    private fun constantTimeEquals(first: String, second: String): Boolean =
        MessageDigest.isEqual(first.toByteArray(), second.toByteArray())
}
